File: examples/delphi/template_method.pas
{$mode objfpc}{$H+}
program template_method;

uses SysUtils;

type
  TDataMiner = class(TObject)
  protected
    function OpenSource(const Path: string): string; virtual; abstract;
    function ExtractData(const Raw: string): string; virtual; abstract;
    function ParseData(const Data: string): string; virtual; abstract;
  public
    procedure Mine(const Path: string);
  end;

  TCSVMiner = class(TDataMiner)
  protected
    function OpenSource(const Path: string): string; override;
    function ExtractData(const Raw: string): string; override;
    function ParseData(const Data: string): string; override;
  end;

  TJSONMiner = class(TDataMiner)
  protected
    function OpenSource(const Path: string): string; override;
    function ExtractData(const Raw: string): string; override;
    function ParseData(const Data: string): string; override;
  end;

  TXMLMiner = class(TDataMiner)
  protected
    function OpenSource(const Path: string): string; override;
    function ExtractData(const Raw: string): string; override;
    function ParseData(const Data: string): string; override;
  end;

procedure TDataMiner.Mine(const Path: string);
var
  Raw, Data, Result_: string;
begin
  Raw := OpenSource(Path);
  Data := ExtractData(Raw);
  Result_ := ParseData(Data);
  WriteLn('  Analysis result: ', Result_);
end;

function TCSVMiner.OpenSource(const Path: string): string;
begin
  WriteLn('  Opening CSV file: ', Path);
  Result := 'name,age\nAlice,30\nBob,25';
end;

function TCSVMiner.ExtractData(const Raw: string): string;
begin
  WriteLn('  Extracting data from CSV rows');
  Result := 'Alice:30, Bob:25';
end;

function TCSVMiner.ParseData(const Data: string): string;
begin
  WriteLn('  Parsing CSV data');
  Result := '2 records found, avg age = 27.5';
end;

function TJSONMiner.OpenSource(const Path: string): string;
begin
  WriteLn('  Opening JSON file: ', Path);
  Result := '{"users":[{"name":"Alice"},{"name":"Bob"}]}';
end;

function TJSONMiner.ExtractData(const Raw: string): string;
begin
  WriteLn('  Extracting data from JSON objects');
  Result := 'Alice, Bob';
end;

function TJSONMiner.ParseData(const Data: string): string;
begin
  WriteLn('  Parsing JSON data');
  Result := '2 users found';
end;

function TXMLMiner.OpenSource(const Path: string): string;
begin
  WriteLn('  Opening XML file: ', Path);
  Result := '<users><user name="Alice"/><user name="Bob"/></users>';
end;

function TXMLMiner.ExtractData(const Raw: string): string;
begin
  WriteLn('  Extracting data from XML nodes');
  Result := 'Alice, Bob';
end;

function TXMLMiner.ParseData(const Data: string): string;
begin
  WriteLn('  Parsing XML data');
  Result := '2 user nodes found';
end;

var
  Miner: TDataMiner;
begin
  WriteLn('=== Template Method Pattern: Data Mining ===');
  WriteLn;

  WriteLn('CSV Mining:');
  Miner := TCSVMiner.Create;
  Miner.Mine('data.csv');
  Miner.Free;
  WriteLn;

  WriteLn('JSON Mining:');
  Miner := TJSONMiner.Create;
  Miner.Mine('data.json');
  Miner.Free;
  WriteLn;

  WriteLn('XML Mining:');
  Miner := TXMLMiner.Create;
  Miner.Mine('data.xml');
  Miner.Free;
end.

File: examples/cpp/template-method.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

class DataMiner {
public:
    virtual ~DataMiner() = default;

    // Template method
    void mine(const std::string& path) {
        std::string raw = openFile(path);
        auto data = extractData(raw);
        auto parsed = parseData(data);
        analyzeData(parsed);
        std::cout << "  Report sent." << std::endl;
    }

protected:
    virtual std::string openFile(const std::string& path) = 0;
    virtual std::string extractData(const std::string& raw) = 0;
    virtual std::vector<std::string> parseData(const std::string& data) = 0;

    void analyzeData(const std::vector<std::string>& data) {
        std::cout << "  Analyzing " << data.size() << " records..." << std::endl;
    }
};

class CSVDataMiner : public DataMiner {
protected:
    std::string openFile(const std::string& path) override {
        std::cout << "  Opening CSV file: " << path << std::endl;
        return "name,age\nAlice,30\nBob,25";
    }
    std::string extractData(const std::string& raw) override {
        std::cout << "  Extracting CSV data" << std::endl;
        return raw;
    }
    std::vector<std::string> parseData(const std::string& data) override {
        std::cout << "  Parsing CSV rows" << std::endl;
        return {"Alice,30", "Bob,25"};
    }
};

class JSONDataMiner : public DataMiner {
protected:
    std::string openFile(const std::string& path) override {
        std::cout << "  Opening JSON file: " << path << std::endl;
        return R"([{"name":"Alice"},{"name":"Bob"}])";
    }
    std::string extractData(const std::string& raw) override {
        std::cout << "  Extracting JSON data" << std::endl;
        return raw;
    }
    std::vector<std::string> parseData(const std::string& data) override {
        std::cout << "  Parsing JSON objects" << std::endl;
        return {"Alice", "Bob"};
    }
};

class XMLDataMiner : public DataMiner {
protected:
    std::string openFile(const std::string& path) override {
        std::cout << "  Opening XML file: " << path << std::endl;
        return "<users><user>Alice</user><user>Bob</user></users>";
    }
    std::string extractData(const std::string& raw) override {
        std::cout << "  Extracting XML data" << std::endl;
        return raw;
    }
    std::vector<std::string> parseData(const std::string& data) override {
        std::cout << "  Parsing XML nodes" << std::endl;
        return {"Alice", "Bob"};
    }
};

int main() {
    std::cout << "=== Template Method: Data Mining ===" << std::endl;

    std::cout << "\nCSV Mining:" << std::endl;
    CSVDataMiner csv;
    csv.mine("data.csv");

    std::cout << "\nJSON Mining:" << std::endl;
    JSONDataMiner json;
    json.mine("data.json");

    std::cout << "\nXML Mining:" << std::endl;
    XMLDataMiner xml;
    xml.mine("data.xml");

    return 0;
}
